### ISSUE TITLE
Python discovery prefers python3 over python3.12

### ISSUE THREAD

**Issue by hynek** [astral-sh/uv#9046]
This is something that happens with uv 0.5.0 and 0.5.1 but not uv 0.4.30.

I've run into this in Ubuntu Noble containers where for some reason my fat build container carries a `/usr/bin/python3` but my slim prod container does not – but it has `/usr/bin/python3.12`. Both are the same system Python 3.12.

Before 0.5.0, if `UV_PYTHON` is set to `python3.12`, uv sync and uv venv would already say `Using CPython 3.12.3 interpreter at: /usr/bin/python3` but the symlink from the venv would go to `/usr/bin/python3.12`. Now it goes to `/usr/bin/python3` and the venv is broken in prod.

The workaround is setting `UV_PYTHON=/usr/bin/python3.12`

**Comment by charliermarsh** [comment]
Yeah this came up in Discord yesterday. I'm not sure what changed exactly. My guess is that we used to fully resolve those symlinks, but now we follow the Python standard library and only use `sys._base_executable` for the virtual environment Python.

**Comment by zanieb** [comment]
Yeah I can't think of anything but #8481 that would have changed here.

**Comment by zanieb** [comment]
I guess the main question is why `sys._base_executable` is wrong here? This should be "broken" with `python -m venv` too right?

**Comment by zanieb** [comment]
I can fix this specific case by preferring a versioned executable over an unversioned one when a version is requested, but it's just more complexity in the discovery rules.

**Comment by charliermarsh** [comment]
@hynek -- Do you have any sort of minimal repro we can use to look at how the Python setup works?

**Comment by zanieb** [comment]
Should be fixed by #9066

**Comment by zanieb** [comment]
I would still like to understand why the base executable path is wrong.

**Comment by bluss** [comment]
Isn't it the case here that there are two docker images. The first python environment has /usr/bin/python. Then the venv is copied into a different image where /usr/bin/python doesn't exist. Uv doesn't do much wrong, the venv was transported between two "incompatible" python environments?

**Comment by charliermarsh** [comment]
Yeah I don't know that the base executable path is "wrong". It just depends on what Python itself reports.

**Comment by zanieb** [comment]
It still seems weird that a different base executable would be reported depending on if `python` or `python3` is queried.

**Comment by sharifhsn** [comment]
Can confirm, closed by #9066

Repro:

```bash
docker run --rm ubuntu:noble bash -euc '
  apt-get update -qq
  apt-get install -qyy -o APT::Install-Recommends=false python3-setuptools python3.12 curl ca-certificates
  curl -LsSf https://astral.sh/uv/0.5.1/install.sh | sh
  export PATH=$HOME/.local/bin:$PATH UV_PYTHON=python3.12 UV_PYTHON_DOWNLOADS=never UV_LINK_MODE=copy
  uv venv /app
  readlink /app/bin/python
  rm -f /usr/bin/python3 /usr/bin/python      # simulate the slim runtime image
  /app/bin/python -V
'
```

uv 0.5.1:

```bash
/usr/bin/python3
bash: line 9: /app/bin/python: No such file or directory
```

uv 0.5.2:

```bash
/usr/bin/python3.12
Python 3.12.3
```

On @zanieb's base-executable question: CPython sets `sys.executable`/`sys._base_executable` to the name the interpreter was launched as and doesn't resolve symlinks, so the same binary reports `/usr/bin/python`, `/usr/bin/python3`, or `/usr/bin/python3.12` depending on how it's invoked. #8481 switched uv to trust that value; since uv was launching via the first matching unversioned name (python3), the venv recorded `/usr/bin/python3`, which then doesn't exist in a slim runtime image that only ships python3.12. #9066 makes uv prefer the request-matching name (python3.12), so CPython reports the versioned path and the venv survives the copy.
